core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ShardsAllocatorModule.java

```diff
@@ -64,8 +65,7 @@ protected void configure() {
             logger.warn("{} allocator has been removed in 2.0 using {} instead", EVEN_SHARD_COUNT_ALLOCATOR_KEY, BALANCED_ALLOCATOR_KEY);
             shardsAllocator = BalancedShardsAllocator.class;
         } else {
-            shardsAllocator = settings.getAsClass(TYPE_KEY, BalancedShardsAllocator.class,
-                    "org.elasticsearch.cluster.routing.allocation.allocator.", "Allocator");
+            throw new IllegalArgumentException("Unknown ShardsAllocator type [" + type + "]");
```

Comment:
we can leave is a TODO  but I think we need a way to plug this in since there are quite some of them out there. I wonder if we can just have a `Map<String, Class<? extends ShardsAllocator>` here and a register method?